src/platform_impl/linux/wayland/clipboard.rs

```diff
@@ -0,0 +1,308 @@
+        }
+    };
+
+    let written = match writer.write((&*content).as_ref()) {
+        Ok(n) if n == (&*content).as_ref().len() => return,
+        Ok(n) => n,
+        Err(err) if err.kind() == ErrorKind::WouldBlock => 0,
+        Err(_) => return,
+    };
+
+    // We weren't able to write all content at once, so add pipe as a `calloop`'s event source
+    // and continue writing when current content will be read.
+    let mut written = Rc::new(written);
+    let left_to_write = content.clone();
+    let writer = Generic::new(writer, Interest::WRITE, Mode::Level);
+    let _ = loop_handle.insert_source(writer, move |_, file, _| {
+        let left_to_write = (&*left_to_write).as_ref();
+        let (n, action) = match file.write(&left_to_write[*written..]) {
+            Ok(n) if *written + n == left_to_write.len() => (n, PostAction::Remove),
+            Ok(n) => (n, PostAction::Continue),
+            Err(err) if err.kind() == ErrorKind::WouldBlock => (0, PostAction::Continue),
+            Err(_) => (0, PostAction::Remove),
+        };
+
+        *Rc::get_mut(&mut written).unwrap() += n;
+        Ok(action)
+    });
+}
+
+/// Handle writing to clipboard's pipes read end.
+fn read_clipboard(
+    reader: ReadPipe,
+    mime: String,
+    metadata: Option<Arc<ClipboardMetadata>>,
+    window_id: WindowId,
+    loop_handle: &LoopHandle<'static, WinitState>,
+) {
+    let reader = unsafe {
+        match raw_fd_into_non_blocking_file(reader.into_raw_fd()) {
+            Ok(reader) => reader,
+            Err(_) => return,
+        }
+    };
+    let mut content = Vec::<u8>::with_capacity(u16::MAX as usize);
+    let mut reader_buffer = [0; u16::MAX as usize];
+    let reader = Generic::new(reader, Interest::READ, Mode::Level);
+    let _ = loop_handle.insert_source(reader, move |_, file, winit_state| {
+        let action = loop {
+            match file.read(&mut reader_buffer) {
+                Ok(0) => {
+                    let event_sink = &mut winit_state.event_sink;
+                    let data = std::mem::take(&mut content);
+                    let content = ClipboardContent {
+                        data,
+                        mime: mime.clone(),
+                        metadata: metadata.clone(),
+                    };
+                    let window_event = WindowEvent::ClipboardContent(content);
+                    event_sink.push_window_event(window_event, window_id);
+                    break PostAction::Remove;
+                }
+                Ok(n) => content.extend_from_slice(&reader_buffer[..n]),
+                Err(err) if err.kind() == ErrorKind::WouldBlock => break PostAction::Continue,
+                Err(_) => break PostAction::Remove,
+            }
+        };
+
+        Ok(action)
+    });
+}
+
+/// Create a `File` from `RawFd` marking it with `O_NONBLOCK`.
+unsafe fn raw_fd_into_non_blocking_file(raw_fd: RawFd) -> Result<File> {
+    let flags = libc::fcntl(raw_fd, libc::F_GETFD);
```

Comment:
I'm surprised this works!

```suggestion
    let flags = libc::fcntl(raw_fd, libc::F_GETFL);
```